### PR TITLE
Feature: Stray Rabbit Tracker

### PR DESCRIPTION
**Pull request by DavidArthurCole**
## What
Add a tracker for stray rabbits, and counting the extra production gained from them.

<details>

image

</details>

## Changelog New Features
+ Added a tracker for stray rabbits caught in the Chocolate Factory. - Daveed

### CODE REVIEW

**Comment by DavidArthurCole** [comment]
image

Progress continues. Still need a couple more patterns before I'm done.

**Comment by DavidArthurCole** [comment]
All patterns are done, this is ready to go.

**Comment by hannibal002** [comment]
Cool idea and good looking code!

Please use `SkyHanniTracker`. See `VerminTracker.kt` for a simple implementation.
A more detailed review will follow later.

**Comment by DavidArthurCole** [comment]
> Cool idea and good looking code!
> 
> Please use `SkyHanniTracker`. See `VerminTracker.kt` for a simple implementation. A more detailed review will follow later.

Unfortunately I won't be able to re-verify all functionality until the next hoppity's event, but looks like the re-implementation using `SkyHanniTracker` is functional.

image

**Review by hannibal002 — CHANGES_REQUESTED** [review]
The implementation looks fine, good job!
One small thing I have noticed:
image
We want to modify both display modes at the same time. IDK anymore why the vermin tracker handles this differently than the other trackers.

**Comment by hannibal002** [comment]
We can extract this behavior the same way we do for `incrementRarity`: (idk the correct function name here)
image

File: src/main/java/at/hannibal2/skyhanni/features/inventory/chocolatefactory/ChocolateFactoryStrayTracker.kt [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+                goldenStrayClick.matchMatcher(loreLine) {
+                    tracker.modify { t -> t.goldenTypesCaught.addOrPut("goldenclick", 1) }
+                }
+
+                // Golden Strays, hoard/stampede
+                if (loreLine == "§7You caught a stray §6§lGolden Rabbit§7! §7A hoard of §aStray Rabbits §7has appeared!") {
+                    tracker.modify { t -> t.goldenTypesCaught.addOrPut("stampede", 1) }
+                }
+
+                //Asynchronously update to immediately reflect caught stray
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/chocolatefactory/ChocolateFactoryStrayTracker.kt]: not necessary. the modify() function already marks the tracker display as dirty, this should cause the display to get redrawn in the next render tick.

**Review by hannibal002 — CHANGES_REQUESTED** [review]

File: src/main/java/at/hannibal2/skyhanni/features/inventory/chocolatefactory/ChocolateFactoryStrayTracker.kt (hunk 1 above)
Comment (DavidArthurCole) [reply src/main/java/at/hannibal2/skyhanni/features/inventory/chocolatefactory/ChocolateFactoryStrayTracker.kt]: Removed in latest commit

File: src/main/java/at/hannibal2/skyhanni/features/inventory/chocolatefactory/ChocolateFactoryStrayTracker.kt [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+            }
+        } catch (e: NoSuchElementException) {
+            return
+        }
+    }
+
+    @SubscribeEvent
+    fun onBackgroundDraw(event: GuiRenderEvent.ChestGuiOverlayRenderEvent) {
+        if (!isEnabled()) return
+        tracker.firstUpdate()
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/chocolatefactory/ChocolateFactoryStrayTracker.kt]: firstUpdate should only get called once. see the other usages of this function

File: src/main/java/at/hannibal2/skyhanni/features/inventory/chocolatefactory/ChocolateFactoryStrayTracker.kt [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+            if (nameText.equals("§6§lGolden Rabbit §8- §aSide Dish")) {
+                claimedStraysSlots.add(event.slot.slotIndex)
+                incrementGoldenType("sidedish")
+                incrementRarity("legendary", 0)
+                DelayedRun.runDelayed(1.seconds) {
+                    claimedStraysSlots.remove(claimedStraysSlots.indexOf(event.slot.slotIndex))
+                }
+            }
+        } catch (e: NoSuchElementException) {
+            return
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/inventory/chocolatefactory/ChocolateFactoryStrayTracker.kt]: this should show an error to the user. we don't want hidden errors.

**Review by hannibal002 — CHANGES_REQUESTED** [review]
works fine in game

File: src/main/java/at/hannibal2/skyhanni/features/inventory/chocolatefactory/ChocolateFactoryStrayTracker.kt (hunk 2 above)
Comment (hannibal002) [reply src/main/java/at/hannibal2/skyhanni/features/inventory/chocolatefactory/ChocolateFactoryStrayTracker.kt]: Maybe on the cf inventory open event?

File: src/main/java/at/hannibal2/skyhanni/features/inventory/chocolatefactory/ChocolateFactoryStrayTracker.kt (hunk 3 above)
Comment (DavidArthurCole) [reply src/main/java/at/hannibal2/skyhanni/features/inventory/chocolatefactory/ChocolateFactoryStrayTracker.kt]: Isn't even really an error - just stack detection. Either way, better method for early-return detection in latest commit.

File: src/main/java/at/hannibal2/skyhanni/features/inventory/chocolatefactory/ChocolateFactoryStrayTracker.kt (hunk 2 above)
Comment (DavidArthurCole) [reply src/main/java/at/hannibal2/skyhanni/features/inventory/chocolatefactory/ChocolateFactoryStrayTracker.kt]: Fixed in latest commit.

**Pull request merged**